velox/type/Timestamp.cpp

```diff
@@ -234,94 +219,166 @@ bool Timestamp::epochToUtc(int64_t epoch, std::tm& tm) {
+
   if (options.precision == TimestampToStringOptions::Precision::kMilliseconds) {
     nanos /= 1'000'000;
   } else if (
       options.precision == TimestampToStringOptions::Precision::kMicroseconds) {
     nanos /= 1'000;
   }
   if (options.skipTrailingZeros && nanos == 0) {
-    return out;
+    return StringView(startPosition, writePosition - startPosition);
   }
-  out += '.';
-  const int offset = out.size();
-  int trailingZeros = 0;
 
+  // Fractional part.
+  *writePosition++ = '.';
+  // Append leading zeros.
+  const auto numDigits = countDigits(nanos);
+  std::memset(writePosition, '0', precisionWidth - numDigits);
+  writePosition += precisionWidth - numDigits;
+
+  // Append the remaining numeric digits.
   if (options.skipTrailingZeros) {
+    std::optional<uint32_t> nonZeroOffset = std::nullopt;
+    int32_t offset = numDigits - 1;
+    // Write non-zero digits from end to start.
     while (nanos > 0) {
-      if (out.size() == offset && nanos % 10 == 0) {
-        trailingZeros += 1;
-      } else {
-        out += '0' + nanos % 10;
+      if (nonZeroOffset.has_value() || nanos % 10 != 0) {
+        *(writePosition + offset) = '0' + nanos % 10;
+        if (!nonZeroOffset.has_value()) {
+          nonZeroOffset = offset;
+        }
       }
+      --offset;
       nanos /= 10;
     }
+    writePosition += nonZeroOffset.value() + 1;
   } else {
-    while (nanos > 0) {
-      out += '0' + nanos % 10;
-      nanos /= 10;
-    }
+    const auto [position, errorCode] =
+        std::to_chars(writePosition, writePosition + numDigits, nanos);
+    VELOX_DCHECK_EQ(
+        errorCode,
+        std::errc(),
+        "Failed to convert fractional part to chars: {}.",
+        std::make_error_code(errorCode).message());
+    writePosition = position;
   }
+  return StringView(startPosition, writePosition - startPosition);
+}
 
-  while (out.size() - offset < precisionWidth - trailingZeros) {
-    out += '0';
-  }
-  std::reverse(out.begin() + offset, out.end());
-  return out;
+StringView Timestamp::tsToStringView(
+    const Timestamp& ts,
+    const TimestampToStringOptions& options,
+    int8_t precisionWidth,
+    char* const startPosition) {
+  std::tm tmValue;
+  VELOX_USER_CHECK(
+      epochToUtc(ts.getSeconds(), tmValue),
+      "Can't convert seconds to time: {}",
+      ts.getSeconds());
+  const uint64_t nanos = ts.getNanos();
+  return tmToStringView(tmValue, nanos, options, precisionWidth, startPosition);
 }
 
 void parseTo(folly::StringPiece in, ::facebook::velox::Timestamp& out) {
   // TODO Implement
 }
 
+std::string::size_type getMaxStringLength(
+    const TimestampToStringOptions& options) {
+  const auto precisionWidth = static_cast<int8_t>(options.precision);
+  switch (options.mode) {
+    case TimestampToStringOptions::Mode::kDateOnly:
+      // Date format is %y-mm-dd, where y has 10 digits at maximum for int32.
+      // Possible sign is considered.
+      return 17;
```

Comment:
In timestamp test, the generated date could be like `-1163612410-02-06` which contains 10 year digits. But in real workload, the maximum year digits should be less. Do we need modify the test?
https://github.com/facebookincubator/velox/blob/82bcd3903ae2f0e27d58b686971958218b3aab55/velox/type/tests/TimestampTest.cpp#L392

Reply:
I note Timestamp's constructor has limited how many seconds it can hold in Timestamp.h. If exceeding, milliseconds in int64 will overflow.

Reply:
Here the test is triggered with a value of `std::tm`, so the timestamp constructor will not be called.